# ML-09 — Validation and Research Claim Audit

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/abc085455-byte/flyrank-ml-internship/blob/main/work/notebooks/w06_validation_audit.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. Two paper findings + my methodology questions

*Pick two findings from the FlyRank research paper. For each: where does the label come from, and does the validation design carry the claim? Constructive tone.*

**Finding #4 — "The Freshness Multiplier" (`docs/flyrank-seo-research-march-2026.pdf`, p.9).**
The paper's most dramatic number: 365+ day content that was refreshed within the last 30 days shows a
3.2x health-score boost (10.7 -> 34.5) and 57x more impressions (71 -> 4,039) compared with old,
un-refreshed content, and the playbook turns this straight into an action item ("Run a recurring refresh
program... produces 3.2x health and 57x impressions").

**My methodology question:** where does the "refreshed" group come from? This reads as an observed
cross-sectional contrast (refreshed old pages vs. non-refreshed old pages), not an experiment — nobody was
randomly assigned to get refreshed. A team is more likely to *spend* refresh effort on pages it still
believes in: ones with retained backlinks, a still-relevant topic, or renewed search demand. If that's true,
part of the 3.2x/57x gap could be "we already picked promising pages to refresh," not "refreshing caused the
boost" — the paper's own evidence standard says observational data doesn't prove causation, and this
finding's action language reads a little further than a single snapshot comparison can carry. A stronger
version would either match refreshed pages to similar-looking non-refreshed pages before comparing, or track
the *same* pages before/after their own refresh date, and it would be even stronger paired with the sample
size behind 71 and 4,039 — the same care the paper already shows next door, where it flags the neighboring
`361+` freshness bucket's 283:1 ratio as unstable because it rests on n=1 declining page.

**Finding — ML Appendix, "What Predicts Growth?" (p.29).**
A logistic regression reaches 71% holdout accuracy separating growing from declining pages across the
61.8K-row ML sample (57 brands), with content age as the strongest negative signal.

**My methodology question:** the Methodology page (p.36) states an 80/20 holdout split for this model but
doesn't say whether the split is grouped by brand. If it's a random row split, pages from the same brand can
land on both sides of the 80/20 divide, and the model can pick up brand-specific quirks (a brand's typical
page age, CMS habits, editorial rhythm) rather than a generally transferable growth signal — inflating the
71% figure the same way a random split inflates ours below. It would strengthen this finding to state
whether the split was grouped by brand, and to print the label's base rate (share of pages actually
growing) next to the 71%, so a reader can see how many of those points are skill versus a majority-class
default — the same "always print the base rate" habit `hunting-leakage-and-validating` asks of us.

Framing note: both of these are asked in the spirit the paper itself sets — it already keeps reversed and
nuanced findings visible on purpose (p.18) — so this is "how to make it stronger," not a scorecard.



In [3]:
# --- Setup: make this notebook work whether it's opened locally or via the Colab badge ---
# Colab's working directory is /content, not the repo -- so relative paths like "docs/..."
# and "data/raw/..." break unless we locate (or clone) the repo root first.
import os, pathlib, subprocess

REPO_URL = "https://github.com/abc085455-byte/flyrank-ml-internship.git"

def find_repo_root(start="."):
    p = pathlib.Path(start).resolve()
    for candidate in [p, *p.parents]:
        if (candidate / "skills" / "README.md").exists() and (candidate / "docs").exists():
            return candidate
    return None

repo_root = find_repo_root()
if repo_root is None:
    clone_dir = pathlib.Path("/content/flyrank-ml-internship")
    if not clone_dir.exists():
        subprocess.run(["git", "clone", REPO_URL, str(clone_dir)], check=True)
    repo_root = clone_dir

os.chdir(repo_root)
print("Working directory set to:", os.getcwd())


Working directory set to: /content/flyrank-ml-internship


## 1. Two paper findings + my methodology questions

*Pick two findings from the FlyRank research paper. For each: where does the label come from, and does the validation design carry the claim? Constructive tone.*

**Finding #4 — "The Freshness Multiplier" (`docs/flyrank-seo-research-march-2026.pdf`, p.9).**
The paper's most dramatic number: 365+ day content that was refreshed within the last 30 days shows a
3.2x health-score boost (10.7 -> 34.5) and 57x more impressions (71 -> 4,039) compared with old,
un-refreshed content, and the playbook turns this straight into an action item ("Run a recurring refresh
program... produces 3.2x health and 57x impressions").

**My methodology question:** where does the "refreshed" group come from? This reads as an observed
cross-sectional contrast (refreshed old pages vs. non-refreshed old pages), not an experiment — nobody was
randomly assigned to get refreshed. A team is more likely to *spend* refresh effort on pages it still
believes in: ones with retained backlinks, a still-relevant topic, or renewed search demand. If that's true,
part of the 3.2x/57x gap could be "we already picked promising pages to refresh," not "refreshing caused the
boost" — the paper's own evidence standard says observational data doesn't prove causation, and this
finding's action language reads a little further than a single snapshot comparison can carry. A stronger
version would either match refreshed pages to similar-looking non-refreshed pages before comparing, or track
the *same* pages before/after their own refresh date, and it would be even stronger paired with the sample
size behind 71 and 4,039 — the same care the paper already shows next door, where it flags the neighboring
`361+` freshness bucket's 283:1 ratio as unstable because it rests on n=1 declining page.

**Finding — ML Appendix, "What Predicts Growth?" (p.29).**
A logistic regression reaches 71% holdout accuracy separating growing from declining pages across the
61.8K-row ML sample (57 brands), with content age as the strongest negative signal.

**My methodology question:** the Methodology page (p.36) states an 80/20 holdout split for this model but
doesn't say whether the split is grouped by brand. If it's a random row split, pages from the same brand can
land on both sides of the 80/20 divide, and the model can pick up brand-specific quirks (a brand's typical
page age, CMS habits, editorial rhythm) rather than a generally transferable growth signal — inflating the
71% figure the same way a random split inflates ours below. It would strengthen this finding to state
whether the split was grouped by brand, and to print the label's base rate (share of pages actually
growing) next to the 71%, so a reader can see how many of those points are skill versus a majority-class
default — the same "always print the base rate" habit `hunting-leakage-and-validating` asks of us.

Framing note: both of these are asked in the spirit the paper itself sets — it already keeps reversed and
nuanced findings visible on purpose (p.18) — so this is "how to make it stronger," not a scorecard.


In [4]:
# Read-only check: confirm the paper is in the repo and pull the page numbers I'm citing above,
# so the methodology questions above are traceable to real pages, not from memory.
import pdfplumber

PAPER_PATH = "docs/flyrank-seo-research-march-2026.pdf"

with pdfplumber.open(PAPER_PATH) as pdf:
    n_pages = len(pdf.pages)
    p9 = pdf.pages[8].extract_text()   # Finding #4 -- Freshness Multiplier
    p29 = pdf.pages[28].extract_text() # ML Appendix -- Growth & Classification
    p36 = pdf.pages[35].extract_text() # Methodology page

print(f"Paper found: {PAPER_PATH} ({n_pages} pages)\n")
print("p.9 contains '3.2x' and '57x':", ("3.2x" in p9) and ("57x" in p9))
print("p.29 contains '71%' holdout accuracy line:", "71%" in p29)
print("p.36 Methodology mentions the ML split design:")
for line in p36.splitlines():
    if "split" in line.lower() or "80/20" in line:
        print("   ", line.strip())


Paper found: docs/flyrank-seo-research-march-2026.pdf (36 pages)

p.9 contains '3.2x' and '57x': True
p.29 contains '71%' holdout accuracy line: True
p.36 Methodology mentions the ML split design:
    (80/20 split), Logistic Regression (80/20 split), PCA (2 components), and a shallow Decision Tree (80/20 split). ML pages remain exploratory


## 2. My model under an honest split (before/after)

*Re-run your Week-5 model under a grouped or time-aware split. Show both numbers.*

Week 5 (`work/notebooks/w05_model.ipynb`) already treats a **client-grouped split** as the honest default
(`GroupShuffleSplit` on `client_id`) rather than a random row split, because `client_id` groups pages that
share a brand, CMS, and editorial team — a random split lets a model see 90% of a client's pages during
training and "solve" its own holdout by memorizing that client's quirks, not by learning anything that
transfers to a client it has never seen.

This section re-runs that exact comparison end to end, on the same features, same label
(`is_declining_label`), same two models (Logistic Regression, Random Forest), so the **before** (naive
random split, stratified on the label) sits next to the **after** (client-grouped split) on every metric —
not just ROC AUC.


In [5]:
import numpy as np
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GroupShuffleSplit, train_test_split
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import (
    roc_auc_score, average_precision_score, precision_score,
    recall_score, f1_score, accuracy_score,
)

RANDOM_STATE = 42
pd.set_option("display.width", 140)

df = pd.read_csv("data/raw/content_refresh_anonymized.csv")
df["is_declining_label"] = (df["trend_direction"] == "down").astype(int)

numeric_features = [
    "log_impressions_90d", "log_clicks_90d", "log_sessions_90d", "log_ai_sessions_90d",
    "days_with_impressions", "days_with_sessions", "content_age_days",
    "days_since_last_update", "ctr", "avg_position", "engagement_rate",
    "scroll_rate", "ai_traffic_pct", "search_volume", "competition", "cpc",
    "word_count", "has_search_volume", "has_word_count",
]
categorical_features = [
    "content_type", "main_intent", "competition_level", "age_tier",
    "freshness_tier", "word_count_tier", "impression_tier", "position_tier",
]
# trend_direction / trend_pct excluded (label source). content_id / client_id excluded
# (pseudonyms, grouping only). No FlyRank product flags exist in this starter file.

df["log_impressions_90d"] = np.log1p(df["impressions_90d"])
df["log_clicks_90d"] = np.log1p(df["clicks_90d"])
df["log_sessions_90d"] = np.log1p(df["sessions_90d"])
df["log_ai_sessions_90d"] = np.log1p(df["ai_sessions_90d"])
df["has_search_volume"] = df["search_volume"].notna().astype(int)
df["has_word_count"] = df["word_count"].notna().astype(int)

num = df[numeric_features].apply(pd.to_numeric, errors="coerce").replace([np.inf, -np.inf], np.nan).fillna(0)
cat = df[categorical_features].fillna("unknown").astype(str)
cat_enc = pd.get_dummies(cat, prefix=categorical_features, dtype=float)
X = pd.concat([num.reset_index(drop=True), cat_enc.reset_index(drop=True)], axis=1)
y = df["is_declining_label"].to_numpy()
groups = df["client_id"].to_numpy()

def precision_at_k(y_true, scores, k):
    frame = pd.DataFrame({"y": y_true, "score": scores})
    top = frame.sort_values("score", ascending=False).head(min(k, len(frame)))
    return float(top["y"].mean())

def full_metrics(y_true, scores, binary_pred):
    return {
        "roc_auc": roc_auc_score(y_true, scores),
        "avg_precision": average_precision_score(y_true, scores),
        "precision_at_50": precision_at_k(y_true, scores, 50),
        "precision@flag": precision_score(y_true, binary_pred, zero_division=0),
        "recall@flag": recall_score(y_true, binary_pred, zero_division=0),
        "f1@flag": f1_score(y_true, binary_pred, zero_division=0),
    }

def fit_score(train_idx, test_idx, model):
    pipe = Pipeline([("scaler", StandardScaler()), ("model", model)]) if isinstance(model, LogisticRegression) else model
    pipe.fit(X.iloc[train_idx], y[train_idx])
    proba = pipe.predict_proba(X.iloc[test_idx])[:, 1]
    return full_metrics(y[test_idx], proba, (proba >= 0.5).astype(int))

# ---------- BEFORE: naive random split, stratified on the label ----------
tr_r, te_r = train_test_split(np.arange(len(X)), test_size=0.2, random_state=RANDOM_STATE, stratify=y)
overlap_r = set(df.iloc[tr_r]["client_id"]) & set(df.iloc[te_r]["client_id"])

lr_before = fit_score(tr_r, te_r, LogisticRegression(class_weight="balanced", max_iter=1000, random_state=RANDOM_STATE))
rf_before = fit_score(tr_r, te_r, RandomForestClassifier(class_weight="balanced_subsample", max_depth=10,
                                                          min_samples_leaf=25, n_estimators=300,
                                                          n_jobs=-1, random_state=RANDOM_STATE))

# ---------- AFTER: honest, client-grouped split ----------
gss = GroupShuffleSplit(n_splits=1, test_size=0.2, random_state=RANDOM_STATE)
tr_g, te_g = next(gss.split(X, y, groups=groups))
overlap_g = set(df.iloc[tr_g]["client_id"]) & set(df.iloc[te_g]["client_id"])

lr_after = fit_score(tr_g, te_g, LogisticRegression(class_weight="balanced", max_iter=1000, random_state=RANDOM_STATE))
rf_after = fit_score(tr_g, te_g, RandomForestClassifier(class_weight="balanced_subsample", max_depth=10,
                                                         min_samples_leaf=25, n_estimators=300,
                                                         n_jobs=-1, random_state=RANDOM_STATE))

print("BEFORE -- naive random split (stratified on label)")
print(f"  client overlap between train/test: {len(overlap_r)} of {df['client_id'].nunique()} total clients")
print(f"  test base rate: {y[te_r].mean():.3f}")
before_table = pd.DataFrame({"logistic_regression": lr_before, "random_forest": rf_before}).T
print(before_table.round(3).to_string())

print("\nAFTER -- honest, client-grouped split")
print(f"  client overlap between train/test: {len(overlap_g)} of {df['client_id'].nunique()} total clients"
      f"  ({df.iloc[tr_g]['client_id'].nunique()} train clients / {df.iloc[te_g]['client_id'].nunique()} test clients)")
print(f"  test base rate: {y[te_g].mean():.3f}")
after_table = pd.DataFrame({"logistic_regression": lr_after, "random_forest": rf_after}).T
print(after_table.round(3).to_string())

print("\nGAP (before minus after, same model / same features / same label)")
gap_table = (before_table - after_table).round(3)
print(gap_table.to_string())


BEFORE -- naive random split (stratified on label)
  client overlap between train/test: 31 of 32 total clients
  test base rate: 0.542
                     roc_auc  avg_precision  precision_at_50  precision@flag  recall@flag  f1@flag
logistic_regression    0.710          0.727             0.92           0.677        0.675    0.676
random_forest          0.757          0.769             0.94           0.707        0.730    0.718

AFTER -- honest, client-grouped split
  client overlap between train/test: 0 of 32 total clients  (25 train clients / 7 test clients)
  test base rate: 0.511
                     roc_auc  avg_precision  precision_at_50  precision@flag  recall@flag  f1@flag
logistic_regression    0.615          0.603             0.72           0.588        0.622    0.605
random_forest          0.607          0.589             0.64           0.588        0.598    0.593

GAP (before minus after, same model / same features / same label)
                     roc_auc  avg_precision  

**Before vs. after, read plainly.** The naive random split leaves 31 of the portfolio's 32 clients present
on *both* sides of train/test — almost every client's pages were partly memorized before being "tested" on.
Under that split, Logistic Regression reaches precision@50 around 0.92 and Random Forest around 0.94. Once
the split is grouped by client (0 client overlap; 7 clients held out entirely), the *same* models on the
*same* features and label drop to roughly 0.72 (Logistic Regression) and 0.64 (Random Forest) —
a 20-30 point fall on the exact metric this lane is read on (precision@50), and a smaller but still real
~0.10-0.15 fall in ROC AUC.

That gap is the finding, not a failure: it's memorization the random split was hiding. The honest, defensible
number for this lane is the grouped-split one — **observed precision@50 of about 0.72 (Logistic Regression)
on 7 held-out clients the model never trained on**, next to a test-set base rate of roughly 0.51 (about half
the queue is genuinely declining either way), so the lift over "flag half the queue at random" is real but
modest, not dramatic.


## 3. Leakage audit

*The same hunt from Week 3, on your final feature set.*

Same attack checklist as `hunting-leakage-and-validating`, run here on the Week-5 final feature set (53
columns after one-hot encoding: 18 numeric + 8 encoded categoricals). Two attacks: (1) deliberately inject a
label-derived column and confirm the score jumps toward 1.0 — if it doesn't, the test harness itself is
broken; (2) drop the single feature that already worried me in Week 5 (`days_with_impressions`, the top
permutation-importance feature for Random Forest) and confirm the model does **not** collapse without it.


In [6]:
from sklearn.linear_model import LogisticRegression as LR2
from sklearn.pipeline import Pipeline as Pipe2
from sklearn.preprocessing import StandardScaler as SS2
from sklearn.metrics import roc_auc_score as auc2

X_train_g, X_test_g = X.iloc[tr_g], X.iloc[te_g]
y_train_g, y_test_g = y[tr_g], y[te_g]

def auc_with(extra_cols):
    Xt, Xte = X_train_g.copy(), X_test_g.copy()
    for c in extra_cols:
        Xt[c] = pd.to_numeric(df.iloc[tr_g][c], errors="coerce").fillna(0).to_numpy()
        Xte[c] = pd.to_numeric(df.iloc[te_g][c], errors="coerce").fillna(0).to_numpy()
    pipe = Pipe2([("scaler", SS2()), ("model", LR2(class_weight="balanced", max_iter=1000, random_state=RANDOM_STATE))])
    pipe.fit(Xt, y_train_g)
    return auc2(y_test_g, pipe.predict_proba(Xte)[:, 1])

honest_auc = auc_with([])
print(f"1) Honest feature set, grouped split, LR ROC AUC:              {honest_auc:.3f}")

leak_pct_auc = auc_with(["trend_pct"])
print(f"   + inject trend_pct (label's own numerator):                {leak_pct_auc:.3f}  <- should jump toward 1.0")

df["_trend_direction_down_copy"] = (df["trend_direction"] == "down").astype(int)  # literal label copy
leak_direct_auc = auc_with(["_trend_direction_down_copy"])
print(f"   + inject a literal copy of the label:                      {leak_direct_auc:.3f}  <- confirms harness is sensitive")

# --- suspect feature: does the top permutation-importance feature collapse the score if removed? ---
feat_minus = [c for c in X_train_g.columns if c != "days_with_impressions"]
lr_minus = Pipe2([("scaler", SS2()), ("model", LR2(class_weight="balanced", max_iter=1000, random_state=RANDOM_STATE))])
lr_minus.fit(X_train_g[feat_minus], y_train_g)
lr_minus_auc = auc2(y_test_g, lr_minus.predict_proba(X_test_g[feat_minus])[:, 1])

rf_full = RandomForestClassifier(class_weight="balanced_subsample", max_depth=10, min_samples_leaf=25,
                                  n_estimators=300, n_jobs=-1, random_state=RANDOM_STATE)
rf_full.fit(X_train_g, y_train_g)
rf_full_auc = auc2(y_test_g, rf_full.predict_proba(X_test_g)[:, 1])

rf_minus = RandomForestClassifier(class_weight="balanced_subsample", max_depth=10, min_samples_leaf=25,
                                   n_estimators=300, n_jobs=-1, random_state=RANDOM_STATE)
rf_minus.fit(X_train_g[feat_minus], y_train_g)
rf_minus_auc = auc2(y_test_g, rf_minus.predict_proba(X_test_g[feat_minus])[:, 1])

print(f"\n2) LR ROC AUC with days_with_impressions:    {honest_auc:.3f}")
print(f"   LR ROC AUC WITHOUT days_with_impressions: {lr_minus_auc:.3f}  <- no meaningful change")
print(f"   RF ROC AUC with days_with_impressions:    {rf_full_auc:.3f}")
print(f"   RF ROC AUC WITHOUT days_with_impressions: {rf_minus_auc:.3f}  <- small drop, not a collapse")

# --- product-flag / decision-derived feature scan ---
flagged_cols = [c for c in df.columns if any(k in c.lower() for k in ["flag", "score", "health", "optimiz"])
                and c != "_trend_direction_down_copy"]
print(f"\n3) Columns in the raw CSV that look like product flags/scores: {flagged_cols if flagged_cols else 'none found'}")

# --- window-overlap check: does impressions_90d (contains the label's own last30/prev30 sub-windows)
#     correlate strongly enough with the label to be a disguised copy of it? ---
corr_imp = np.corrcoef(np.log1p(df["impressions_90d"]), df["is_declining_label"])[0, 1]
corr_days = np.corrcoef(df["days_with_impressions"], df["is_declining_label"])[0, 1]
print(f"\n4) corr(log1p(impressions_90d), label) = {corr_imp:.3f}   corr(days_with_impressions, label) = {corr_days:.3f}"
      "   (both weak -- nowhere near a disguised copy)")

df.drop(columns=["_trend_direction_down_copy"], inplace=True)


1) Honest feature set, grouped split, LR ROC AUC:              0.615
   + inject trend_pct (label's own numerator):                0.999  <- should jump toward 1.0
   + inject a literal copy of the label:                      1.000  <- confirms harness is sensitive

2) LR ROC AUC with days_with_impressions:    0.615
   LR ROC AUC WITHOUT days_with_impressions: 0.615  <- no meaningful change
   RF ROC AUC with days_with_impressions:    0.607
   RF ROC AUC WITHOUT days_with_impressions: 0.593  <- small drop, not a collapse

3) Columns in the raw CSV that look like product flags/scores: none found

4) corr(log1p(impressions_90d), label) = 0.177   corr(days_with_impressions, label) = 0.190   (both weak -- nowhere near a disguised copy)


**Verdict.** The harness works and the top feature is not a leak.

1. **The confession test passes.** Injecting `trend_pct` (the label's own numerator) lifts LR ROC AUC from
   ~0.62 to ~1.00; injecting a literal copy of `trend_direction` hits 1.000 exactly — the near-perfect
   collapse-toward-one the leakage skill predicts for a true label sibling, which confirms this test harness
   is sensitive enough to trust. Neither column is in the real feature set — both are excluded per
   `docs/data-dictionary.md`'s explicit label-source rule.

2. **`days_with_impressions` is a real feature, not a disguised label.** Removing it changes Logistic
   Regression's AUC not at all, and costs Random Forest only ~0.01-0.02 AUC. A true leak collapses toward
   the base rate when removed; this doesn't. It makes sense as the top driver on its own terms: a page going
   quiet on more of the last 90 days is, almost by definition, a page whose visibility is fading.

3. **No decision-derived / product-flag features exist in this starter file** — confirmed by scanning
   column names for `flag`/`score`/`health`/`optimiz` and finding nothing (the only match was the copy I
   injected on purpose for test 1). `provider_used` / `model_used` are excluded per the data dictionary's
   "not a model feature" note.

4. **Window overlap — a disclosed nuance, not a leak.** `impressions_90d` (and everything summed over the
   90-day window) technically *contains* the `last_30d` / `prev_30d` sub-windows the label is computed from.
   That would be a real problem for a *forward-looking* forecast, but this task (per `w02_ml_task_framing`)
   is deliberately framed as same-period triage — "is this page currently declining?" — not "will it decline
   next?" — so features drawn from the same 90-day window as the label are expected, not hidden. The weak
   correlations above (0.18-0.19, nowhere near the ~1.0 a real leak would show) back that up. Worth
   restating as a limitation, though: this model cannot be read as predicting *future* decline, only as
   scoring *current* decline risk from data that already contains it.

5. **IDs used only for grouping** — `content_id` / `client_id` never enter the feature matrix, only
   `groups=` in the split.


## 4. Claim rewrite

*Take my own boldest sentence and rewrite it in safe language: observed, measured, directional, decision-support.*

**My boldest sentence (from `outputs/model_report.md`, echoed in `work/notebooks/w02_ml_task_framing.ipynb`
Section 5):**

> "the random forest scores 0.740 precision@50 (about 37 of 50 real)"

This states a single, precise number as if it were a fixed property of the model. Section 2 above already
shows that's too confident in one direction — a naive random split reports 0.94, not 0.74, for the same
model. But even the *honest*, client-grouped number turns out to be less stable than a single sentence
implies: with only 32 clients in this starter slice, *which* 7 end up in the held-out test set changes the
grouped-split precision@50 meaningfully.


In [7]:
from sklearn.model_selection import GroupShuffleSplit as GSS3
from sklearn.ensemble import RandomForestClassifier as RFC3

seed_results = []
for seed in [42, 1, 7, 13, 99]:
    gss3 = GSS3(n_splits=1, test_size=0.2, random_state=seed)
    tr3, te3 = next(gss3.split(X, y, groups=groups))
    rf3 = RFC3(class_weight="balanced_subsample", max_depth=10, min_samples_leaf=25,
               n_estimators=300, n_jobs=-1, random_state=RANDOM_STATE)
    rf3.fit(X.iloc[tr3], y[tr3])
    proba3 = rf3.predict_proba(X.iloc[te3])[:, 1]
    seed_results.append({
        "split_seed": seed,
        "test_clients": df.iloc[te3]["client_id"].nunique(),
        "precision_at_50": round(precision_at_k(y[te3], proba3, 50), 3),
        "roc_auc": round(roc_auc_score(y[te3], proba3), 3),
    })

seed_table = pd.DataFrame(seed_results)
print(seed_table.to_string(index=False))
print(f"\nprecision@50 across 5 client-grouped splits: min={seed_table['precision_at_50'].min():.2f}, "
      f"max={seed_table['precision_at_50'].max():.2f}, mean={seed_table['precision_at_50'].mean():.2f}")


 split_seed  test_clients  precision_at_50  roc_auc
         42             7             0.64    0.607
          1             7             0.84    0.641
          7             7             0.78    0.667
         13             7             0.88    0.668
         99             7             0.74    0.630

precision@50 across 5 client-grouped splits: min=0.64, max=0.88, mean=0.78


**Rewritten claim:**

> Across five different client-grouped holdouts of the same 32-client starter dataset, Random Forest's
> observed precision@50 ranged from about 0.64 to 0.88 (mean ~0.77), compared with a naive random-split
> reading of 0.94 on the same model and features. **This is decision-support evidence for a specific,
> narrow claim: on this 30K-row starter sample, a client-grouped Random Forest model directionally
> outperforms flagging half the queue at random, and its precision@50 is best reported as a range, not a
> single decimal, given how few clients (32) this dataset holds.** It does not support "the model achieves
> 0.740 precision@50" as a fixed capability, and it does not yet support any claim about performance on
> brand-new clients outside this sample, since all five seeds still draw their 7 test clients from the same
> pool of 32.

What changed, in plain terms: "the model scores 0.740" (stated once, sounds exact) became "the model's
observed precision@50 lands in a 0.64-0.88 range across resampled honest splits, with a real (if smaller)
edge over the base rate" (measured, ranged, and scoped to *this* dataset and *this* validation design).


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.